components/api_request_helper/api_request_helper.cc

```diff
@@ -92,9 +116,25 @@ void APIRequestHelper::OnResponse(
       }
     }
   }
+
   url_loaders_.erase(iter);
-  std::move(callback).Run(response_code, response_body ? *response_body : "",
-                          headers);
+  if (!response_body) {
+    std::move(callback).Run(response_code, "", headers);
+    return;
+  }
+
+  auto raw_body = *response_body;
+  if (conversion_callback) {
+    auto converted_body = std::move(conversion_callback).Run(raw_body);
+    if (converted_body) {
+      raw_body = converted_body.value();
```

Comment:
I thought the callback is intended for conversion only, not for canceling requests, @bridiver wdyt? added a test cases for this

Reply:
I think we should keep this way because:
1. when we have lots of requests (this is quite high loaded part) copying the body for each one can affect the performance, would be better to have the option to skip the conversion without copying the response body, it could be big
2. Errors like wrong responses should be handled on consumer’s side, here we just call a functor to make some preparation steps for our json parser. Consumers can always add own value (or change the response to be error structured) to handle it on the consumer's side.

Reply:
If the conversion fails then we should error here, not wait until it tries to parse an invalid integer because it didn't get converted to a string. We should inform the caller here what the error was so we have the correct error and not a less useful error down the line.

Reply:
I think we also have a problem here because we're only returning http status codes so we don't have any good way to indicate a json conversion/parsing error here. I think for now we can do something like return 422 along with the parsing error message.

Reply:
but in general we want to start using absl::Optional and I'm going to create a PR to make a wrapper for it https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++-features.md#Status-allowed

Reply:
`If the conversion fails then we should error here` the conversion callbacks written by the consumers, in this case they can decide what to do with the answer, return an empty string in the response or overwrite it as `{ error: "conversion failed", "code": 422}` and handle it

Reply:
but they don't know it failed because we're not notifying them, we're just sending the raw body without the conversion. If `converted_body` is null we need to respond here with an error, not return the unconverted raw body

Reply:
```
if (converted_body) {
  raw_body = converted_body.value();
} else {
  std::move(callback).Run(422, raw_body, headers);
}
```

Reply:
done

Reply:
nit: 
```
if (!converted_body) {
  std::move(callback).Run(422, raw_body, headers);
  return;
}
raw_body = converted_body.value();
```

Reply:
done